x/incentive/genesis.go

```diff
@@ -34,18 +36,105 @@ func NewGenesisState(
 	}
 }
 
-// DefaultGenesisState returns the default genesis state
-func DefaultGenesisState() *GenesisState {
+// DefaultGenesis returns the default genesis state
+func DefaultGenesis() *GenesisState {
 	return &GenesisState{
 		Params:          DefaultParams(),
 		NextProgramId:   1,
 		LastRewardsTime: 0,
 	}
 }
 
-// ValidateGenesis checks a genesis state for basic issues
-func ValidateGenesis(_ GenesisState) error {
-	// TODO #1749
+// Validate checks a genesis state for basic issues
+func (gs GenesisState) Validate() error {
+	if err := gs.Params.Validate(); err != nil {
+		return err
+	}
+	if gs.NextProgramId == 0 {
+		return ErrInvalidProgramID.Wrap("next program ID was zero")
+	}
+	if gs.LastRewardsTime < 0 {
+		return ErrDecreaseLastRewardTime.Wrap("last reward time was negative")
+	}
+	// TODO: enforce no duplicate (account,denom)
+	for _, rt := range gs.RewardTrackers {
+		if _, err := sdk.AccAddressFromBech32(rt.Account); err != nil {
+			return err
+		}
+		if !leveragetypes.HasUTokenPrefix(rt.UToken) {
+			return leveragetypes.ErrNotUToken.Wrap(rt.UToken)
+		}
+		if err := rt.Rewards.Validate(); err != nil {
+			return err
+		}
+		for _, r := range rt.Rewards {
+			if leveragetypes.HasUTokenPrefix(r.Denom) {
+				return leveragetypes.ErrUToken.Wrap(r.Denom)
+			}
+		}
```

Comment:
can we move this loop to `Rewards.Validate()`?

Reply:
Since that is just a call to `sdk.DecCoins.Validate()`, we cannot

Reply:
New validate functions solved this though